### ISSUE TITLE
why privileged: true

### ISSUE THREAD

**Issue by trondhindenes** [OctopusDeploy/helm-charts#24]
I'm noticing that the statefulset runs with `privileged: true` without explaining the reasoning for it, or allowing users to set it to false. I would expect that to be documented if it's a hard requirement for Octopus to work

**Comment by MJRichardson** [comment]
@trondhindenes `priviledged: true` is to allow docker-in-docker, for running [execution containers ](https://octopus.com/docs/projects/steps/execution-containers-for-workers)on the [built-in worker](https://octopus.com/docs/infrastructure/workers#built-in-worker). 

It can be disabled by setting [enabledDockerInDocker](https://github.com/OctopusDeploy/helm-charts/blob/main/charts/octopus-deploy/values.yaml#L41C12-L41C12) to `false`, e.g.

```
octopus:
  enableDockerInDocker: false
```

Perhaps this should be false by default?  

We're working on some changes to make it much easier to run Octopus tasks on a k8s cluster.  Once these are done, I think we may disable the built-in worker for self-hosted k8s installations.

There is a [comment in the values file](https://github.com/OctopusDeploy/helm-charts/blob/main/charts/octopus-deploy/values.yaml#L37).  Do you think there is a better place to document this?

**Comment by todaywasawesome** [comment]
@MJRichardson is this still required now that we have the Kubernetes agent?

**Comment by CarlosZambonii** [comment]
Short answer: it depends on which chart/pod you mean. `privileged: true` is **still required** for the legacy `octopus-deploy` (built-in worker) chart when Docker-in-Docker is enabled, but the **Kubernetes Agent's own pod does not require it** — it doesn't hardcode `privileged` at all. There is, however, one unrelated and still-unconditional `privileged: true` in the `kubernetes-agent` chart's NFS StatefulSet that has nothing to do with DinD and isn't gated by any value.

**(a) `octopus-deploy` chart (built-in worker) — privileged is still tied to `enableDockerInDocker`, default `true`**
- `templates/statefulset.yaml` only injects `securityContext.privileged: true` on the `octopus` container when `.Values.octopus.enableDockerInDocker` is truthy (it's appended after any user-supplied `containerSecurityContext`).
- `values.yaml` defaults `octopus.enableDockerInDocker: true` and comments explain this is because the built-in worker's execution-containers feature needs Docker-in-Docker, which requires the Server container to run privileged.
- `README.md` explicitly tells non-root/strict-security-context users to set `enableDockerInDocker: false`, and says it "must" be `false` for read-only-root-filesystem setups.
- Confirmed by test: `tests/values/security_context.yaml` sets `enableDockerInDocker: false`, and `tests/od_statefulset_test.yaml` asserts the rendered `securityContext` is exactly `{privileged: false, readOnlyRootFilesystem: false, runAsNonRoot: true, runAsUser: 999, runAsGroup: 999}` with no `privileged: true` injected.

So for this chart, the answer to "is this still required" is: only if you keep DinD-based execution containers on the built-in worker; it can be turned off today via `enableDockerInDocker: false`.

**(b) `kubernetes-agent` chart — the agent pod itself is not privileged**
- `templates/tentacle-deployment.yaml` sets the pod-level `securityContext` entirely from `.Values.agent.securityContext` (defaults to `{}`), with no `privileged` key set by the template — `grep -n privileged` on this file returns nothing.
- `templates/pre-install-tentacle-pod.yaml` has the identical pattern (`.Values.agent.securityContext`, default `{}`), again no hardcoded `privileged`.
- `templates/script-pod-template.yaml` (the per-execution script pods that replace DinD in this model) is built purely from user-configurable `scriptPods.podTemplate.*` values, also defaulting to `{}` — no hardcoded `privileged` there either.
- `charts/kubernetes-agent/README.md` never mentions Docker-in-Docker or `enableDockerInDocker` at all; instead each task execution gets its own script pod (`scriptPods.*` values), which is the architectural reason DinD/privileged isn't needed here.
- The `crds/script-pod-template.yaml` CRD schema does declare a `privileged` boolean field, but that's standard PodSpec schema boilerplate for the CR (its own description says "Defaults to false") — not a hardcoded `true`.

**(c) One other unconditional `privileged: true` remains in `kubernetes-agent`, unrelated to DinD**
- `templates/nfs-statefulset.yaml` hardcodes `privileged: true` on the NFS server container (lines ~39-40) with no `{{ if }}`/`{{ with }}` gating and no corresponding `.Values` key — a repo-wide grep confirms there's no `persistence.nfs.securityContext` or `persistence.nfs.privileged` value to override it (the only override point, `persistence.nfs.containers.nfs.spec`, is rendered before this literal block, so `privileged: true` always wins regardless).
- I found no evidence in this repository (issue, comment, changelog, or code comment) that this NFS privileged usage is tracked or planned for removal — I'd treat that as a separate, still-open item rather than assume it's already being handled elsewhere.

---
**TL;DR:** For the built-in worker (`octopus-deploy` chart), yes — `privileged: true` is still needed when `enableDockerInDocker` is `true` (the default), and can be disabled via that flag. For the Kubernetes Agent (`kubernetes-agent` chart), no — the agent pod (`templates/tentacle-deployment.yaml`) and script pods (`templates/script-pod-template.yaml`) don't hardcode `privileged` at all. The one remaining unconditional `privileged: true` in the Kubernetes Agent chart is unrelated to DinD — it's in `templates/nfs-statefulset.yaml` for the NFS server container, has no override value, and doesn't appear to be tracked anywhere yet.
